### PR TITLE
feat: created collection and item creators

### PR DESCRIPTION
**Pull request by Ben-Cheng**
Fixes: #43

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -14,16 +14,14 @@ jobs:
...
           pip install -r requirements-dev.txt
+          pip install -r requirements.txt
       - name: Lint Last Commit
         if: github.event_name == 'push'
         run: |
           gitlint
       - name: Format
         run: |
           black . --check --diff
-      - name: Lint
```
Comment (blacha) [inline .github/workflows/ci.yml:24]: Should add linting back in

File: historical_aerial_photos_metadata.csv [hunk 2]
```diff
@@ -0,0 +1,4 @@
+WKT,sufi,survey,run,photo_no,alternate_survey_name,camera,camera_sequence_no,nominal_focal_length,altitude,scale,photocentre_lat,photocentre_lon,date,film,film_sequence_no,photo_type,format,source,physical_film_condition,image_anomalies,scanned,raw_filename,released_filename,when_scanned,photo_version
```
Comment (blacha) [inline historical_aerial_photos_metadata.csv:1]: test data should not live at the top level

File: main.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import os
+
+import pystac as stac
+
+from topo_processor.metadata import create_collection
+
+tiff_file_path = os.path.join(os.getcwd(), "test_data")
```
Comment (blacha) [inline main.py:7]: should revisit the names of these variables, a collection is not a tiff_file

File: topo_processor/metadata/collection.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    license: str
+    data_type: DataType
+    items: List
+    providers: List
+
+    def __init__(self, title: str, description: str, licence: str, data_type: str):
+        self.title = title
+        self.description = description
+        self.license = licence
+        self.data_type = DataType(data_type)
```
Comment (blacha) [inline topo_processor/metadata/collection.py:26]: pass this in as a `DataType` and do the validation before the constructor.

File: topo_processor/metadata/collection.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+class Collection:
+    """
+    collection of items
+    """
+
+    tilte: str
+    description: str
+    license: str
+    data_type: DataType
+    items: List
```
Comment (blacha) [inline topo_processor/metadata/collection.py:19]: `List` needs a type as to what is inside the list.

File: topo_processor/metadata/collection_test.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+from topo_processor.metadata import Collection, DataType
+
+
+def test_collection_has_title():
```
Comment (blacha) [inline topo_processor/metadata/collection_test.py:4]: I am not sure how important these tests are, you are only validating the constructor actually assigns something.

if anything you should have one test that does all the validation of the constructor

File: topo_processor/metadata/item.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        self.stac_item = stac.Item(
+            stac_extensions=[],
+            id="hash",  # TODO
+            geometry=None,  # TODO
+            bbox=None,  # TODO
+            properties={},
+            datetime=datetime.now(),
+        )
+
+        # # TODO: create thumbnail
```
Comment (blacha) [inline topo_processor/metadata/item.py:23]: Creating thumbnails should not be done here, as a Item may not actually have a thumbnail to be created.

File: topo_processor/metadata/loaders/metadata_loader_imagery_historic.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    def is_applicable(self, item: Item) -> bool:
+        if item.collection.data_type != DataType.ImageryHistoric:
+            return False
+        if not is_tiff(item.path):
+            return False
+        return True
+
+    def add_metadata(self, item: Item):
+        file_name = os.path.splitext(os.path.basename(item.path))[0]  # nicer way
```
Comment (blacha) [inline topo_processor/metadata/loaders/metadata_loader_imagery_historic.py:22]: `# nicer way`? might want to expand on what you mean by that

File: topo_processor/metadata/loaders/metadata_loader_imagery_historic.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        csv_path = os.path.join(os.getcwd(), "historical_aerial_photos_metadata.csv")
+
+        if not os.path.isfile(csv_path):
+            raise Exception('Missing "historical_aerial_photos_metadata.csv"')
+
+        with open(csv_path, "r") as fd:
+            csv_file = csv.reader(fd, delimiter=",")
+            # reader = csv.reader(csv_file)
```
Comment (blacha) [inline topo_processor/metadata/loaders/metadata_loader_imagery_historic.py:30]: this line is not needed

File: topo_processor/metadata/collection.py (hunk 5 above)
Comment (Ben-Cheng) [reply topo_processor/metadata/collection.py:19]: That will create a circular import again where collection.py needs to import item (And we are already importing collection in item.py). Does that mean we need an item factory?

File: topo_processor/metadata/loaders/metadata_loader_imagery_historic.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            file_name_index = metadata_keys.index("released_filename")
+
+            has_found_file = False
+            for row in csv_file:
+                if file_name != row[file_name_index]:
+                    continue
+                if has_found_file:
+                    raise Exception(f'Duplicate file "{file_name}" found in metadata csv')
+                has_found_file = True
+                properties = {
```
[the remark this answers is not in the document]
Comment (blacha) [reply topo_processor/metadata/loaders/metadata_loader_imagery_historic.py:40]: These are not going to be the end metadata points as they are literally just the database field names.

I need to add a ticket to review the metadata extensions to STAC so we can look to standardise how we do extensions across Basemaps, Datalake and Processor

File: topo_processor/metadata/loaders/metadata_loader_repo.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+from topo_processor.metadata.item import Item
+from topo_processor.util.time import time_in_ms
+
+from .metadata_loader import MetadataLoader
+
+
+class MetadataLoaderRepository:
+    loaders: List[MetadataLoader] = []
+
+    def append_loader(self, loader: MetadataLoader) -> None:
```
Comment (blacha) [inline topo_processor/metadata/loaders/metadata_loader_repo.py:14]: you generally dont add the name of the thing to the repository, as the repository is already telling you its a  "MetadataLoader"

so you could just do `append`

File: topo_processor/metadata/loaders/metadata_loader_imagery_historic.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (blacha) [reply topo_processor/metadata/loaders/metadata_loader_imagery_historic.py:40]: Ticket made: https://github.com/linz/topo-processor/issues/56

[the remark this answers is not in the document]
Comment (palmerj) [reply topo_processor/metadata/loaders/metadata_loader_imagery_historic.py:40]: Great.

**Review by blacha — APPROVED** [review]

**Pull request merged**